### PR TITLE
Specify how to handle message as payload 

### PR DESCRIPTION
**Pull request by zolkis**
Handle #357

***
<a href="https://pr-preview.s3.amazonaws.com/zolkis/web-nfc/pull/359.html" title="Last updated on Sep 30, 2019, 3:29 PM UTC (1d342d3)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/web-nfc/359/748e682...zolkis:1d342d3.html" title="Last updated on Sep 30, 2019, 3:29 PM UTC (1d342d3)">Diff</a>

### CODE REVIEW

**Review by kenchris — APPROVED** [review]
Would like a code example though

**Comment by kenchris** [comment]
Want to merge or are you adding a code example first?

**Comment by zolkis** [comment]
Wait, I will add examples (wrongly discussed in the issue, whereas I meant to comment here).

**Comment by zolkis** [comment]
@kenchris it's better to review commit by commit, as much of the changes are related to making examples sub-sections.
Then, let's merge and do the rest in other PRs.

**Comment by zolkis** [comment]
I am thinking to better specify smart poster for instance: now it's relying on the SP RTD spec (e.g. actions are 0, 1 and 2, with no context explained).

File: index.html [hunk 1]
```diff
@@ -919,6 +919,93 @@ <h4>
   reader.scan({ recordType: "example.com:sp"});
       </pre>
     </aside>
+    <aside title="Push a smart poster message" class="example">
+      <p>
+        Smart poster defined in NFC Forum Smart Poster Record Type Definition
```
Comment (kenchris) [inline index.html]: In this example we write a Smart Poster as defined in...

File: index.html [hunk 2]
```diff
@@ -919,6 +919,93 @@ <h4>
...
+        specification. It is an NDEF record that contains an <a>NDEF message</a>
+        as payload, which in turn may contain several records: a URI record that
```
Comment (kenchris) [inline index.html]: which at least contains the following records (need to check spec, maybe only URL is required)

File: index.html [hunk 3]
```diff
@@ -919,6 +919,93 @@ <h4>
...
+        refers to a content, and additional optional records for that content's
+        title, icon, size, type and possible action (e.g. send SMS, make call,
+        launch browser, or save for later, or open for editing, etc).
+      </p>
+      <pre class="highlight">
+        const writer = new NFCWriter();
+        let sp = new NDEFMessage({
```
Comment (kenchris) [inline index.html]: you are not modifying so use const

File: index.html [hunk 4]
```diff
@@ -919,6 +919,93 @@ <h4>
...
+          records: [
+            {
+              recordType: "url",
+              data: "https://my.org/content/19911"
+            },
+            {
+              recordType: "t", // type, a local type to Sp
```
Comment (kenchris) [inline index.html]: maybe explain what it represents? Maybe we should make a non-nominative section quickly explaining what a smart poster may contain?

Comment (kenchris) [reply index.html]: ah you do explain what it means, `type` just made it confusing.

"smart poster type, a local type to Sp"

File: index.html [hunk 5]
```diff
@@ -919,6 +919,93 @@ <h4>
...
+            {
+              recordType: "act",  // action, a local type to Sp
+              data: 0  // do the action, in this case open in the browser
+            },
+          ];
+        });
+        writer.push({ records: [
+          {
+            recordType: "smart-poster",
+            data: sp.records
```
Comment (kenchris) [inline index.html]: If you are just calling `sp.records` then why do we need to create the NDEFMessage object and not just an array of json records?

File: index.html [hunk 6]
```diff
@@ -919,6 +919,93 @@ <h4>
...
+      </pre>
+    </aside>
+    <aside title="Push an external record with an NDEF message as payload"
+           class="example">
+      <p>
+        External type records can be used to create application defined records
+        that may even contain an <a>NDEF message</a> as payload.
+      </p>
+      <p>
+        This example show how to do the previous Smart Poster use case with
```
Comment (kenchris) [inline index.html]: I find it confusing to refer to Smart Poster here as you are not storing anythign like a smart poster

**Comment by kenchris** [comment]
> I am thinking to better specify smart poster for instance: now it's relying on the SP RTD spec (e.g. actions are 0, 1 and 2, with no context explained).

Not sure what you mean.

File: index.html [hunk 7]
```diff
@@ -624,113 +624,111 @@ <h4>
...
-const writer = new NFCWriter();
-writer.push(
-  "Hello World"
-).then(() => {
-  console.log("Message pushed.");
-}).catch(error => {
-  console.log(`Push failed :-( try again: ${error}.`);
-});
+    <pre class="example">
+      const writer = new NFCWriter();
```
Comment (kenchris) [inline index.html:634]: Doesn't this extra indentation add extra indentation to the spec?

**Comment by kenchris** [comment]
> > I am thinking to better specify smart poster for instance: now it's relying on the SP RTD spec (e.g. actions are 0, 1 and 2, with no context explained).
> 
> Not sure what you mean.

I don't want to create a special smart poster object or so, because they can contain other records, like the Signature spec adds a new record to it. But we could specify it but maybe as that doesn't require any implementation it is basically just a non-nominative informative section. That is what I was suggesting

File: index.html (hunk 7 above)
Comment (zolkis) [reply index.html:634]: No, I compared it with the original, no difference at the rendered doc.

File: index.html (hunk 5 above)
Comment (zolkis) [reply index.html]: I thought is was more clear as an example. You could have that Sp message from other source as well (e.g. previously read). But no problem, I can compact the example.

File: index.html (hunk 6 above)
Comment (zolkis) [reply index.html]: OK, I can skip that remark. What I meant is that you can share the same content in 2 ways: with the standardized Sp, or in an app-specific way using external records and embedded NDEF message.

File: index.html (hunk 2 above)
Comment (zolkis) [reply index.html]: Yes, only URL is mandatory. We might want to standardize Sp in Web NFC indeed.

Comment (kenchris) [reply index.html]: More add an informal section describing it (as we were standardizing it, more or less)

Comment (zolkis) [reply index.html]: Let's handle the Sp in a next PR. I made the suggested changes, please re-check.

File: index.html [hunk 8]
```diff
@@ -871,54 +868,132 @@ <h4>
...
+        writer.push({ records: [
+          {
+            recordType: "smart-poster",
+            data: { records: [
+              {
+                recordType: "url",
+                data: "https://my.org/content/19911"
+              },
+              {
+                recordType: "t", // smart poster type, a local type to Sp
```
Comment (kenchris) [inline index.html:941]: How are these local types exposed when reading the records? `smart-poster:t` ?

Anyway we can look at that in follow ups

**Review by kenchris — APPROVED** [review]

File: index.html (hunk 8 above)
Comment (zolkis) [reply index.html:941]: According to the Sp RTD it's just the "t" for title and "act" for action, then the regular tnf=1 text, tnf=1 uri, and tnf=2 (mime type).

**Comment by zolkis** [comment]
Merging for now and let's take the Sp informative section, plus other changes in next PRs.

File: index.html (hunk 8 above)
Comment (kenchris) [reply index.html:941]: yes but these are local, so that t is restricted to Sp records, but I guess that is fine exposing it like that

**Pull request merged**